src/docs/binding-overview.md

```diff
@@ -0,0 +1,105 @@
+- Swift has top level entities, i.e., global functions and variables.
+- In Swift, all members in a type *must* be initialized. Similarly, unsafe pointer types have a notion of 4 possible states that they can be in:
+  - allocated
+  - initialized
+  - deinitialized
+  - deallocated
+Effectively, all types follow this pattern, but the 4 states are managed by the compiler and the runtime and unlike pointer types, there is no explicit control of it.
+
+## ABI Differences
+There are a number of changes that Apple made outside of the standard ABIs used for target CPUs that are important to keep in mind:
+- For value types, they allow more registers to be used for passing the type.
+- For instance methods and closures, there is a separate register reserved outside of the regular argument registers used for passing the instance pointer
+- For functions that throw, there is a separate register reserved for exceptions (errors)
+- For returning value types an extra register is available
+
+## Runtime Differences
+The main difference in the Swift runtime is that heap allocated types are reference counted rather than using another memory management scheme. The reference counting is not done with a single count but instead is managed with two counts: a strong count and a weak count. Weak references are available in Swift for making (potentially) circular data structures.
+
+## Idiomatic Differences
+For the most part, these are minor. C# prefers camel case starting with a capital letter for all types and entry points, using lowercase starts for private fields and argument names. C# adopts that conventions that interfaces start with 'I'. Swift starts types with capitals but public variables will start with lowercase.
+
+Swift favors value types over heap allocated types and favors copy-on-write semantics.
+
+For bit fields, C# prefers using enums with the `[Flags]` attribute. In Swift, the pattern is to use a struct that implements `OptionSet`.
+
+## Accessibility
+First and foremost, our customers matter the most. They are the ones who will be using the code and everything they use should "feel" like C# to them rather than Swift. As much as possible, this should be automated.
+
+That said, there will be hard choices that need to be made. For example, consider the following Swift function:
+```swift
+public func VowelCount(text: String) -> Int { /* ... */ }
+```
+`String` in Swift is a vastly different and completely incompatible type from a C# string. In C# a string is an array of UTF16 characters. In Swift, the layout of `String` is completely different. It is a union of several different possible representations. As such, it is always the case that when you want to make a Swift string accessible in C# you will need to allocate memory. Similarly, when you want a C# string accessible in Swift, you will need to allocate memory. Consequently, you have to care about the memory lifetime of the object.
```

Comment:
For this particular issue, I strong recommend looking at how COM interop handled these sorts of issues. It seems to me that adding similar APIs for this sort of interop is completely appropriate [`Marshal.GetComObjectData()`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getcomobjectdata) and would make the ergonomics work nicely for both the user and how this could integrate with our custom marshallers. I could imagine a new `SwiftStringMarshaller` that handles all of this. Since C# `string` types are immutable this approach works, but I do see how other scenarios involving mutable types may not. Something to consider.

/cc @jkoritzinsky

Reply:
One thing to have in mind that this will run on platforms where both the trimmer and AOT are necessary. That makes any kind of dynamic implementation a lot trickier to implement.

Reply:
There is nothing dynamic about the type information though. This is merely associating already allocated typed memory with an already allocated type instance. All the types are statically known and so I'm unclear how this specific `string` 
 example would be difficult to enable for trimming. If we were storing some dynamic type, I get it, but I don't see how this approach would impact trimming.

Reply:
That's great!

I merely wanted to bring trimming/AOT up in case it could cause a problem (I have no idea how the COM interop is implemented, but the [method](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getcomobjectdata?view=net-8.0) you linked to doesn't look strongly typed at all - which is very often a sign something isn't trimmable), because it's better to catch these problems early on. But if it's all statically known, then no problem whatsoever.